### DISCUSSION TITLE
useSuspenseQuery - throw error while having stale data

### DISCUSSION THREAD

**Discussion by grzesiek-ds** [TanStack/query#6960] [Q&A] [answered]
Hello TanStack team!

First of all I would like to thank You for Your hard work - TanStack Query is a great tool and I enjoy using it.

## Example situation
Consider a case where we have component which is suppose to always display fresh (not-stale) data.

- When data is loading for first time, loader should be displayed.
- When fresh data is available, it should be displayed.
- When we did not manage to fetch fresh data, and data in cache becomes stale or theres no data in cache, error boundary should be displayed.
- When error boundary is resetting, loader should be displayed (since data in cache is stale, or theres no data in cache).

## Problem statement
I believe that it could be achieved by using `useQuery` and custom `throwOnError`, but while using `useSuspenseQuery`, it appears that custom `useEffect` is needed.

### With `useQuery`
```javascript
// Note - I did not test it!
const QueryConsumer = () => {
  const result = useQuery({
    ...,
    staleTime: 30 * 1000,
    throwOnError: (_, query) => typeof query.state.data === 'undefined' || query.isStaleByTime(30 * 1000),
  })
}
```

### With `usesSuspenseQuery`
```javascript
// Also did not test it!
const QueryConsumer = () => {
  const result = useQuery({
    ...,
  })
  // useLayoutEffect instead to throw error before rendering?
  useEffect(() => {
    if (result.isStale && result.error) {
      throw result.error;
    }
  }, [result.isStale, result.error]);
}
```

## Question
Since `throwOnError` is not available for `useSuspenseQuery`.
https://tanstack.com/query/latest/docs/framework/react/guides/suspense#throwonerror-default

Would You consider allowing `throwOnError` for `useSuspenseQuery` while also preserving current `defaultThrowOnError` logic?

### Current implementation
```javascript
const defaultThrowOnError = (...) => typeof query.state.data === 'undefined'

const useSuspenseQuery = (...) => {
  return useBaseQuery({
    ...,
    throwOnError: defaultThrowOnError,
  })
}

const useBaseQuery = (...) => {
  ...
  if (getHasError(...)) {
    throw result.error
  }
  ...
}

export const getHasError = (...) => {
  return (
    result.isError &&
    !errorResetBoundary.isReset() &&
    !result.isFetching &&
    query &&
    shouldThrowError(throwOnError, [result.error, query])
  )
}
```

### Proposed change
```javascript
// useSuspenseQuery can define throwOnError, which is combined with defaultThrownError
const useSuspenseQuery = ({ ..., throwOnError }) => {
  return useBaseQuery({
    ...,
    throwOnError: (error, query) => defaultThrowOnError(error, query) || throwOnError?.(error, query)
  })
}

// Or defaultThrowOnError logic is defined in getHasError - this affects useQuery behaviour
export const getHasError = (...) => {
  return (
    result.isError &&
    !errorResetBoundary.isReset() &&
    !result.isFetching &&
    query &&
    (
      // Logic from defaultThrowOnError is defined in getHasError
      typeof query.state.data === 'undefined' ||
      // Custom, optional throwOnError is allowed
      shouldThrowError(throwOnError, [result.error, query])
    )
  )
}
```

**Comment by TkDodo** [comment]
`throwOnError` is not allowed on `useSuspenseQuery` because you could set it to `() => false`, and that would void the guarantee that `data` is always defined after `useSuspenseQuery` has been called.

What you can do is just throw during rendering, without needing a useEffect:

```
if (result.isStale && result.error) {
  throw result.error;
}
```

that should trigger the error boundary. It's also in the docs: https://tanstack.com/query/latest/docs/framework/react/guides/suspense#throwonerror-default

**Reply by grzesiek-ds** [reply]
@TkDodo I think that I didnt phrase it correctly - sorry. 
My proposition is to merge the current behaviour of `defaultThrowOnError` with custom value of `throwOnError`.

Therefore when settings `throwOnError: () => false` in reality check would be: `typeof query.state.data === 'undefined' || false`.

As result it would be ensured that data is defined.

**Reply by TkDodo** [reply]
> My proposition is to merge the current behaviour of defaultThrowOnError with custom value of throwOnError.

interesting idea. Wouldn't it be quite confusing to pass `throwOnError: () => false` but still get errors thrown to boundaries if they are for an initial fetch? It would also be a different behaviour of the same option between suspense and non-suspense hooks. Is the throw-in-an-if-statement approach not good enough ?

**Reply by grzesiek-ds** [reply]
I agree that defining `throwOnError` for a suspense query is not great, as the error on the initial fetch cannot be ignored.
Would it be desirable to introduce such functionality as a separate prop, with a more self-explanatory name, just for `useSuspenseQuery`? For example, `throwOnSubsequentError`?
I think that good documentation would be enough to explain this — I'm willing to contribute to both the change and documentation.

There's nothing wrong with the custom error handling you mentioned above.
My goal was to propose the "normalization" of features offered by both `useQuery` and `useSuspenseQuery`. Of course, I'm motivated by a specific case that I have 🙃.

**Reply by TkDodo** [reply, marked as answer]
imo normalization of features only works if we have them behave the same, which we just agreed we cannot do.

**Reply by grzesiek-ds** [reply]
So the answer is "no we don't want this"? 😄

**Reply by TkDodo** [reply]
right now I'd say that's right :)

**Reply by MrBr** [reply]
Maybe the time has change now!

My "default" expectations were that any error in suspense `queryFn` would throw. My expectations could be wrong, but I think things have indeed change from how query was used in the past.

Primarily, now we have `queryOptions` that help encapsulate a single query configuration. Typically (at least how I see it), a query should either throw or not, it doesn't seem "per use case" configuration. So I'd like to configure that option in one place and not think about it in multiple places when used.

Secondly, I'm almost exclusively using `useSuspenseQuery`, this mean that I'm not using query `state` to handle different query "lifecycle" state changes, they're handled automatically with composition. By this I'm referring to `loading` (now handled with Suspense boundary) and `error` (now handled with Error boundary) which we're used very often with `useQuery`. So accessing query state with suspense is more an exception than the rule, and doing it brings up certain questions.

With the old pattern (without suspense) using state from query made a lot of sense, this was the way, and very often every query had a thin hook wrapper that'd encapsulate the specific configuration. Today, this pattern is "outdated" and I think at least the option to always throw with `useSuspenseQuery` is something that would align how we build applications today.

**Reply by TkDodo** [reply]
the problem is if you always throw, you have no way of opting out of it if you want to prefer stale data. The flow I’m talking about is:

- fetch data, success
- work with the app
- get a background refetch
- refetch fails

now we _could_ continue to show stale data, or we show an error boundary. If we show stale data per default, you can still throw yourself for things where it’s absolutely necessary. If we throw per default, users have no way to opt out of it.

what we could do is allow `throwOnError` for suspense, but only with the type `throwOnError?: true`. no function, no false, and keep the current default behaviour ?

**Reply by MrBr** [reply]
This would solve my problem since I'd like to throw always, however I'm not sure if it's the best way. In my case, (stale) following requests got 401 status code after the token expiration. I don't want to handle 401 specifically ad-hoc, I want to handle that logic in one place with Error Boundary, that's why I want for all 401 request to throw and bubble up. In this case it's also fine that all other error throws, but I can see a case where I'd be fine with stale data as well for the same query and a different error.

Food for though. If I had to bet, I'd bet that devs in average add error boundary, because it's the nature of suspense, and more often than not, do not handle following errors. To me this means that a certain states of applications aren't handled purposefully to be stale but rather stay stale unintentionally. Is stale better than forcing devs to think about recovery? I don't know, but it's for sure worth thoughts.
